### PR TITLE
Response codes refactor

### PR DESCRIPTION
**Pull request by kelvinator07**

### CODE REVIEW

File: src/main/java/co/rsk/federate/signing/hsm/client/HSMResponseCodeEnum.java [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+
+    private static final HashMap<Integer, HSMResponseCodeEnum> RESPONSE_CODES = new HashMap<>();
+
+    static {
+        for (HSMResponseCodeEnum responseCodeEnum : values()) {
+            RESPONSE_CODES.put(responseCodeEnum.responseCode, responseCodeEnum);
+        }
+    }
+
+    private final int responseCode;
```
Comment (marcos-iov) [inline src/main/java/co/rsk/federate/signing/hsm/client/HSMResponseCodeEnum.java:37]: Variable declaration should go before methods

File: src/main/java/co/rsk/federate/signing/hsm/client/HSMResponseCodeEnum.java [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+
+    HSMResponseCodeEnum(int responseCode) {
+        this.responseCode = responseCode;
+    }
+
+    public static HSMResponseCodeEnum valueOfResponseCode(int responseCode) throws HSMDeviceException {
```
Comment (marcos-iov) [inline src/main/java/co/rsk/federate/signing/hsm/client/HSMResponseCodeEnum.java:43]:
```suggestion
    public static HSMResponseCodeEnum valueOf(int responseCode) throws HSMDeviceException {
```
It's already implied that it is a response code

File: src/test/java/co/rsk/federate/signing/hsm/client/HSMClientProtocolTest.java [hunk 3]
```diff
@@ -175,7 +175,7 @@ public void sendInvalidVersion() throws JsonRpcException  {
         ObjectNode expectedRequest = new ObjectMapper().createObjectNode();
         expectedRequest.put("command", command);
         expectedRequest.put("version", version);
-        ObjectNode sendResponse = buildResponse(-666);
+        ObjectNode sendResponse = buildResponse(-904);
```
Comment (marcos-iov) [inline src/test/java/co/rsk/federate/signing/hsm/client/HSMClientProtocolTest.java:178]: Why this change?

File: src/test/java/co/rsk/federate/signing/hsm/client/HSMResponseHandlerV1Test.java [hunk 4]
```diff
@@ -36,7 +36,7 @@ public void createResponseHandler() {
     @Test(expected = HSMAuthException.class)
     public void validateResponseRejectedKeyError() throws HSMClientException {
         ObjectNode response = new ObjectMapper().createObjectNode();
-        response.put("errorcode", -1);
+        response.put("errorcode", -103);
```
Comment (marcos-iov) [inline src/test/java/co/rsk/federate/signing/hsm/client/HSMResponseHandlerV1Test.java:39]: Did the response codes change for HSM v1?

File: src/main/java/co/rsk/federate/signing/hsm/client/HSMResponseCodeEnum.java [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+    HSM_TIP_MISMATCH_ERROR_CODE(-203),
+    HSM_INVALID_OR_NOT_ENOUGH_INPUT_BLOCKS_ERROR_CODE(-204),
+    HSM_INVALID_BROTHERS_ERROR_CODE(-205),
+    HSM_INVALID_USER_DEFINED_VALUE_ERROR_CODE(-301),
+    HSM_FORMAT_ERROR_CODE(-901),
+    HSM_INVALID_REQUEST_ERROR_CODE(-902),
+    HSM_COMMAND_UNKNOWN_ERROR_CODE(-903),
+    HSM_VERSION_CHANGED_ERROR_CODE(-904),
+    HSM_DEVICE_ERROR_CODE_V2(-905),
+    HSM_UNKNOWN_ERROR_CODE(-906);
```
[the remark this answers is not in the document]
Comment (kelvinator07) [reply src/main/java/co/rsk/federate/signing/hsm/client/HSMResponseCodeEnum.java:27]: Resolved

File: src/main/java/co/rsk/federate/signing/hsm/client/HSMResponseCodeEnum.java [hunk 6]
```diff
@@ -0,0 +1,49 @@
+package co.rsk.federate.signing.hsm.client;
+
+import co.rsk.federate.signing.hsm.HSMDeviceException;
+
+import java.util.HashMap;
+
+/**
+ * Created by Kelvin Isievwore on 04/05/2023.
+ */
+public enum HSMResponseCodeEnum {
```
[the remark this answers is not in the document]
Comment (kelvinator07) [reply src/main/java/co/rsk/federate/signing/hsm/client/HSMResponseCodeEnum.java:10]: Resolved

File: src/main/java/co/rsk/federate/signing/hsm/client/HSMResponseCodeEnum.java (hunk 1 above)
Comment (kelvinator07) [reply src/main/java/co/rsk/federate/signing/hsm/client/HSMResponseCodeEnum.java:37]: Resolved

File: src/main/java/co/rsk/federate/signing/hsm/client/HSMResponseCodeEnum.java [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+    HSM_INVALID_BROTHERS_ERROR_CODE(-205),
+    HSM_INVALID_USER_DEFINED_VALUE_ERROR_CODE(-301),
+    HSM_FORMAT_ERROR_CODE(-901),
+    HSM_INVALID_REQUEST_ERROR_CODE(-902),
+    HSM_COMMAND_UNKNOWN_ERROR_CODE(-903),
+    HSM_VERSION_CHANGED_ERROR_CODE(-904),
+    HSM_DEVICE_ERROR_CODE_V2(-905),
+    HSM_UNKNOWN_ERROR_CODE(-906);
+
+    private static final HashMap<Integer, HSMResponseCodeEnum> RESPONSE_CODES = new HashMap<>();
```
[the remark this answers is not in the document]
Comment (kelvinator07) [reply src/main/java/co/rsk/federate/signing/hsm/client/HSMResponseCodeEnum.java:29]: Resolved

File: src/main/java/co/rsk/federate/signing/hsm/client/HSMResponseCodeEnum.java (hunk 2 above)
Comment (kelvinator07) [reply src/main/java/co/rsk/federate/signing/hsm/client/HSMResponseCodeEnum.java:43]: Resolved

File: src/main/java/co/rsk/federate/signing/hsm/client/HSMResponseCodeEnum.java [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+
+    HSMResponseCodeEnum(int responseCode) {
+        this.responseCode = responseCode;
+    }
+
+    public static HSMResponseCodeEnum valueOfResponseCode(int responseCode) throws HSMDeviceException {
+        if (RESPONSE_CODES.get(responseCode) == null) {
+            throw new HSMDeviceException("Invalid HSM response code type", responseCode);
+        }
+        return RESPONSE_CODES.get(responseCode);
```
[the remark this answers is not in the document]
Comment (kelvinator07) [reply src/main/java/co/rsk/federate/signing/hsm/client/HSMResponseCodeEnum.java:47]: Resolved

File: src/test/java/co/rsk/federate/signing/hsm/client/HSMClientProtocolTest.java (hunk 3 above)
Comment (kelvinator07) [reply src/test/java/co/rsk/federate/signing/hsm/client/HSMClientProtocolTest.java:178]: -666 Is No Longer The Code For That Particular Error, It Wasn't Updated All This While

File: src/test/java/co/rsk/federate/signing/hsm/client/HSMResponseHandlerV1Test.java (hunk 4 above)
Comment (kelvinator07) [reply src/test/java/co/rsk/federate/signing/hsm/client/HSMResponseHandlerV1Test.java:39]: Yes, -666 Is No Longer The Code For That Particular Error, It Wasn't Updated All This While

File: src/main/java/co/rsk/federate/signing/hsm/client/HSMResponseCodeEnum.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (marcos-iov) [reply src/main/java/co/rsk/federate/signing/hsm/client/HSMResponseCodeEnum.java]: Same with the ERROR_CODE postfix, what do you think?

**Comment by marcos-iov** [comment]
See the code smell reported by Sonar

File: src/main/java/co/rsk/federate/signing/hsm/client/HSMResponseCodeEnum.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (kelvinator07) [reply src/main/java/co/rsk/federate/signing/hsm/client/HSMResponseCodeEnum.java]: Removing The Post Fix Might Not Make Some of The Error Names Look Meaningful

File: src/main/java/co/rsk/federate/signing/hsm/client/HSMResponseCode.java [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+import co.rsk.federate.signing.hsm.HSMDeviceException;
+
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * Created by Kelvin Isievwore on 04/05/2023.
+ */
+public enum HSMResponseCode {
+    DEVICE_ERROR_CODE_V1(-2),
```
Comment (marcos-iov) [inline src/main/java/co/rsk/federate/signing/hsm/client/HSMResponseCode.java:12]:
```suggestion
    V1_DEVICE_ERROR(-2),
```

File: src/main/java/co/rsk/federate/signing/hsm/client/HSMResponseCode.java [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+    SERVER_ERROR_CODE(-4),
```
Comment (marcos-iov) [inline src/main/java/co/rsk/federate/signing/hsm/client/HSMResponseCode.java:13]:
```suggestion
    SERVER_ERROR(-4),
```

File: src/main/java/co/rsk/federate/signing/hsm/client/HSMResponseCode.java [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+    WRONG_AUTH_ERROR_CODE(-101),
+    INVALID_MESSAGE_ERROR_CODE(-102),
+    REJECTED_KEY_ERROR_CODE(-103),
+    CHAINING_MISMATCH_ERROR_CODE(-201),
```
Comment (marcos-iov) [inline src/main/java/co/rsk/federate/signing/hsm/client/HSMResponseCode.java:17]: Let's remove ERROR_CODE from these ones. The name already implies it's an error

File: src/main/java/co/rsk/federate/signing/hsm/client/HSMResponseCode.java [hunk 12]
```diff
@@ -0,0 +1,50 @@
...
+    POW_VALIDATION_ERROR_CODE(-202),
```
Comment (marcos-iov) [inline src/main/java/co/rsk/federate/signing/hsm/client/HSMResponseCode.java:18]:
```suggestion
    POW_VALIDATION_ERROR(-202),
```

File: src/main/java/co/rsk/federate/signing/hsm/client/HSMResponseCode.java [hunk 13]
```diff
@@ -0,0 +1,50 @@
...
+    TIP_MISMATCH_ERROR_CODE(-203),
+    INVALID_OR_NOT_ENOUGH_INPUT_BLOCKS_ERROR_CODE(-204),
+    INVALID_BROTHERS_ERROR_CODE(-205),
+    INVALID_USER_DEFINED_VALUE_ERROR_CODE(-301),
```
Comment (marcos-iov) [inline src/main/java/co/rsk/federate/signing/hsm/client/HSMResponseCode.java:22]: Let's remove ERROR_CODE from these ones. The name already implies it's an error

File: src/main/java/co/rsk/federate/signing/hsm/client/HSMResponseCode.java [hunk 14]
```diff
@@ -0,0 +1,50 @@
...
+    FORMAT_ERROR_CODE(-901),
+    INVALID_REQUEST_ERROR_CODE(-902),
+    COMMAND_UNKNOWN_ERROR_CODE(-903),
+    VERSION_CHANGED_ERROR_CODE(-904),
```
Comment (marcos-iov) [inline src/main/java/co/rsk/federate/signing/hsm/client/HSMResponseCode.java:26]: Let's remove ERROR_CODE from these ones. The name already implies it's an error

File: src/main/java/co/rsk/federate/signing/hsm/client/HSMResponseCode.java [hunk 15]
```diff
@@ -0,0 +1,50 @@
...
+    DEVICE_ERROR_CODE_V2(-905),
```
Comment (marcos-iov) [inline src/main/java/co/rsk/federate/signing/hsm/client/HSMResponseCode.java:27]:
```suggestion
    V2_DEVICE_ERROR(-905),
```

File: src/main/java/co/rsk/federate/signing/hsm/client/HSMResponseCode.java [hunk 16]
```diff
@@ -0,0 +1,50 @@
...
+    UNKNOWN_ERROR_CODE(-906);
```
Comment (marcos-iov) [inline src/main/java/co/rsk/federate/signing/hsm/client/HSMResponseCode.java:28]:
```suggestion
    UNKNOWN_ERROR(-906);
```

File: src/main/java/co/rsk/federate/signing/hsm/client/HSMResponseCode.java [hunk 17]
```diff
@@ -0,0 +1,51 @@
...
+
+import co.rsk.federate.signing.hsm.HSMDeviceException;
+
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * Created by Kelvin Isievwore on 04/05/2023.
+ */
+public enum HSMResponseCode {
```
Comment (marcos-iov) [inline src/main/java/co/rsk/federate/signing/hsm/client/HSMResponseCode.java:11]: Aren't we missing codes 0 and 1 here? There's a method `getOkErrorCodes` in HSMResponseHandlerBase that would make use of them. Also in PowHSMResponseHandler

File: src/main/java/co/rsk/federate/signing/hsm/client/HSMResponseCodeEnum.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (kelvinator07) [reply src/main/java/co/rsk/federate/signing/hsm/client/HSMResponseCodeEnum.java]: Resolved

File: src/test/java/co/rsk/federate/signing/hsm/client/HSMResponseHandlerV1Test.java (hunk 4 above)
Comment (kelvinator07) [reply src/test/java/co/rsk/federate/signing/hsm/client/HSMResponseHandlerV1Test.java]: Resolved

File: src/main/java/co/rsk/federate/signing/hsm/client/HSMResponseCode.java (hunk 17 above)
Comment (kelvinator07) [reply src/main/java/co/rsk/federate/signing/hsm/client/HSMResponseCode.java:11]: 0 And 1 Are Success Codes, Not Error Codes, The Method `getOkErrorCodes` Already Handles The Success Path

Comment (marcos-iov) [reply src/main/java/co/rsk/federate/signing/hsm/client/HSMResponseCode.java:11]: Yeah, and the enum is for HSM response codes not just error codes. I think it makes sense to have them all in one place. getOkErrorCodes can then fetch them from the enum instead of having them hardcoded

File: src/main/java/co/rsk/federate/signing/hsm/client/HSMResponseHandlerBase.java [hunk 18]
```diff
@@ -25,15 +25,17 @@
...
 public class HSMResponseHandlerBase {
     protected static final String ERROR_CODE_FIELD = "errorcode";
-    private static final int HSM_DEVICE_ERROR_CODE_V1 = -2;
-    private static final int HSM_DEVICE_ERROR_CODE_V2 = -905;
 
     //When a `getVersion` request is sent to HSM a `device error` can occur
     //Since the HSM version is yet undefined we need to handle error codes for both versions at this stage
     protected void handleErrorResponse(String methodName, int errorCode, JsonNode response) throws HSMClientException {
-        if (errorCode == HSM_DEVICE_ERROR_CODE_V1 || errorCode == HSM_DEVICE_ERROR_CODE_V2) {
+        HSMResponseCode errorCodeEnum = HSMResponseCode.valueOf(errorCode);
```
[the remark this answers is not in the document]
Comment (marcos-iov) [reply src/main/java/co/rsk/federate/signing/hsm/client/HSMResponseHandlerBase.java:37]: Did you forget to push this change?

**Review by marcos-iov — APPROVED** [review]
